python/src/iceberg/io/base.py

```diff
@@ -0,0 +1,121 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from abc import ABC, abstractmethod
+from typing import Optional
+
+
+class InputFile(ABC):
+    """A base class for InputFile implementations"""
+
+    def __init__(self, location: str):
+        self._location = location
+
+    @abstractmethod
+    def __len__(self) -> Optional[int]:
```

Comment:
Why optional? If the file doesn't exist, should it throw an exception instead? Or is that not a good pattern for Python?

Reply:
I don't like this Optional hint here either. mypy is a bit confused since `self._content_length` is set to `None` when the class is initialized so it complains about just the `int` type hint which is actually accurate. It can't reason about that there's no logic that would return `None` here. I'll just add in a `# type: ignore` comment here and change it to `int`.

Reply:
Updated it to `int` and added `# type: ignore` to the return line for the `__len__` methods to avoid the following from the mypy linter:
```
error: Incompatible return value type (got "Optional[int]", expected "int"
```